Create icarGroupEventResource to support event observations on groups

cookeac opened the pull request:
Defines `icarInventoryClassificationType`, `icarAnimalSetReferenceType`, `icarGroupEventMethodType`.
Extends `icarEventCoreResource` to `icarGroupEventCoreResource`.

Resolves #273 
Resolves #291

Commit:
Create icarInventoryClassification.json

Inventory classification class used to represent livestock that are not individually tagged by their shared characteristics

Commit:
Merge branch 'Develop' of https://github.com/cookeac/ICAR into Develop

Commit:
Define icarGroupEventResource

icarGroupEventResource allows recording of events on groups of animals.
Resolves #273

Commit:
Correct name of icarInventoryClassificationType

Rename from icarInventoryClassification.

Commit:
Rename icarGroupEventMethodType

Renamed from icarGroupEventMethod

AndreasSchultzGEA commented on types/icarAnimalSetReferenceType.json:
```diff
@@ -0,0 +1,16 @@
+        { "$ref": "../types/icarResourceReferenceType.json" },
+        {
+            "type": "object",
+            "properties": {
```
no required properties?

cookeac replied:
Good point. ID should be I think.

cookeac replied:
But see also #248 which might be a more generalised solution.

cookeac replied:
Changed to align with #248 and made `identifier` required.

AndreasSchultzGEA commented on types/icarInventoryClassificationType.json:
```diff
@@ -0,0 +1,46 @@
+      "type": "string",
+      "description": "Human-readable name for this inventory grouping."
+    },
+    "count": {
```
does it clash with same property in icarGroupEventCoreResource?

cookeac replied:
My feeling is the count in the event is the number of animals sampled (if not all the animals were observed) whereas here its the number of animals in the group.
I've modified `icarGroupEventCoreResource`, changing `count` to `countObserved` for clarity.

AndreasSchultzGEA commented on types/icarInventoryClassificationType.json:
```diff
@@ -0,0 +1,46 @@
+  "type": "object",
+  "description": "This type is used to categorise animals by shared characteristics - so you can say the equivalent of 200 x 2-year-old in-calf Jersey heifers.",
+
+  "properties": {
```
nothing required?

cookeac replied:
We want to allow users to specify as much or as little as required, but I did wonder if for instance `name` and `species` should be made mandatory. As this is an embedded element, is it likely that some implementers would not be able to leave it out entirely and then be forced to fill in these fields even if using an animal set instead?

erwinspeybroeck replied:
Always difficult to really say what is required - mostly it depends on the data providers capabilities and the data consumer needs ...
I would say name would be sufficient?

cookeac replied:
Making name and id required.

AndreasSchultzGEA commented on resources/icarGroupEventCoreResource.json:
```diff
@@ -0,0 +1,35 @@
+    {
+      "type": "object",
+
+      "properties": {
```
nothing required?

cookeac replied:
Its a base class - I think required fields should be specified in concrete classes.

erwinspeybroeck replied:
agree

Commit:
Improvements to animal set reference, group event

Make `animalSetId` required in `icarAnimalSetReferenceType`.
Renamce `count` to `countObserved` in `icarGroupEventCoreResource`.

cookeac committed:
Merge branch 'adewg:Develop' into Develop

cookeac committed:
Update PR based on comments.

Address comments in the PR:
- Add fields to icarResourceReferenceType as per #248
- Adjust `icarAnimalSetReferenceType` to be a specialised `icarResourceReferenceType` and make `identifier` required.
- Make `name` and `species` required in `icarInventoryClassificationType`.

cookeac committed:
Merge branch 'Develop' of https://github.com/cookeac/ICAR into Develop

AndreasSchultzGEA commented on resources/icarAnimalEventCoreResource.json:
```diff
@@ -1,5 +1,5 @@
-  "description": "",
+  "description": "Extends icarEventCoreResource to add a single individual animal identifier.",
 
   "allOf": [
```
"animal" as required?

cookeac replied:
Happy to do this though better to set required fields in concrete implementation classes.

AndreasSchultzGEA commented on resources/icarGroupEventCoreResource.json:
```diff
@@ -0,0 +1,35 @@
+      "$ref": "../resources/icarEventCoreResource.json"
+    },
+    {
+      "type": "object",
```
"groupMethod" as required?
Or isn't at least one groupMethod required?

cookeac replied:
Agreed, added.

AndreasSchultzGEA commented on types/icarAnimalSetReferenceType.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "description": "References an animal set through its unique ID in the source system, optionally also specifying its URL (@ID).",
+
+    "required": [ "identifier" ],
```
identifier is not present

cookeac replied:
OK. `identifier` was added to icarResourceReferenceType.json. I didn't want to make it required there though, as this would be a breaking change for other users of that type - medicine and device references. Let me know if this causes code generation or implementation problems. Otherwise at the very least it is documentation that we want identifier to be specified.

cookeac committed:
Make groupMethod required.

cookeac committed:
Made animal required.
